File: src/leetcode/tree/common/tree-node.ts
// This class definition comes from the problem itself and we cannot change it, or else our submission will not be
// accepted.
export class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;

  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

// The LeetCode test input is given as an array, but the exact code to construct the tree is not given.  This is likely
// how they've implemented it.
export function array2tree(array: (number | null)[]): TreeNode | null {
  if (array.length === 0) {
    return null;
  }

  const root = new TreeNode(array[0]!);
  const queue = [root];
  let i = 1;

  while (queue.length > 0 && i < array.length) {
    const node = queue.shift()!;
    if (node === null) {
      continue;
    }

    // Add the left node.
    if (i < array.length) {
      const val = array[i];
      if (val !== null) {
        node.left = new TreeNode(val);
        queue.push(node.left);
      } else {
        node.left = null;
      }
      i++;
    }

    // Add the right node.
    if (i < array.length) {
      const val = array[i];
      if (val !== null) {
        node.right = new TreeNode(val);
        queue.push(node.right);
      } else {
        node.right = null;
      }
      i++;
    }
  }

  return root;
}

File: src/leetcode/tree/common/tree_node.py
from collections import deque


# This class definition comes from the problem itself and we cannot change it, or else our submission will not be
# accepted.
class TreeNode:
    def __init__(self, val=0, left: "TreeNode | None" = None, right: "TreeNode | None" = None):
        self.val = val
        self.left = left
        self.right = right


def array2tree(array: list[int | None]) -> TreeNode | None:
    if not array:
        return None

    if not array[0]:
        return None

    root = TreeNode(array[0])
    queue = deque([root])
    i = 1

    while queue and i < len(array):
        node = queue.popleft()

        if not node:
            continue

        if i < len(array):
            val = array[i]
            if val is not None:
                node.left = TreeNode(val)
                queue.append(node.left)
            i += 1

        if i < len(array):
            val = array[i]
            if val is not None:
                node.right = TreeNode(val)
                queue.append(node.right)
            i += 1

    return root
